lib/primitive/lib/Cardano/Wallet/Primitive/Types/MetadataEncryption.hs

```diff
@@ -0,0 +1,436 @@
+            -> Either ErrMetadataEncryption [(TxMetadataValue, TxMetadataValue)]
+        encryptPairIfQualifies = \case
+            (TxMetaText "msg", m) -> do
+                bimap ErrCannotEncryptMetadata toPair (encryptValue m)
+            pair ->
+                Right [pair]
+
+        encryptValue :: TxMetadataValue -> Either CipherError ByteString
+        encryptValue
+            = AES256CBC.encrypt WithPadding secretKey iv saltM
+            . BL.toStrict
+            . Aeson.encode
+            . metadataValueToJsonNoSchema
+
+        toPair :: ByteString -> [(TxMetadataValue, TxMetadataValue)]
+        toPair encryptedMessage =
+            [ (TxMetaText cip83EncryptPayloadKey, TxMetaList (toChunks encryptedMessage))
+            , (TxMetaText cip83EncryptMethodKey, TxMetaText cip83EncryptPayloadValue)
+            ]
+
+        toChunks :: ByteString -> [TxMetadataValue]
+        toChunks
+            = fmap TxMetaText
+            . T.chunksOf 64
+            . T.decodeUtf8
+            . convertToBase Base64
+
+    updateTxMetadata :: TxMetadataValue -> TxMetadata
+    updateTxMetadata v = TxMetadata (Map.insert cip20MetadataKey v themap)
+      where
+        TxMetadata themap = payload
+
+data ErrMetadataDecryption =
+      ErrMissingMetadataKey
+    | ErrMissingEncryptionMethod
+    | ErrMissingValidEncryptionPayload
+    | ErrCannotAesonDecodePayload Text
+    | ErrMissingSalt
+    | ErrCannotDecryptPayload CipherError
+    | ErrEncryptedPayloadWrongBase
+    deriving (Show, Eq)
+
+-- When decryption is enabled we do the following:
+-- (a) retrieve TxMetaMap under proper key, ie.674,
+--     cip20MetadataKey
+-- (b) check if there is ("enc", "basic") pair
+-- (c) recreate each encrypted payload from chunks that are expected under proper key, ie.msg,
+--     cip83EncryptPayloadKey. So
+--     expect TxMetaList [TxMetaText chunk1, ..., TxMetaText chunkN]
+--     and construct payload=chunk1+chunk2+...+chunkN
+-- (d) decrypt payload and decode metadata
+-- (e) update structure under msg key and remove ("enc", "basic") pair
+fromMetadataEncrypted
+    :: ByteString
+    -> TxMetadata
+    -> Either ErrMetadataDecryption TxMetadata
+fromMetadataEncrypted pwd metadata =
+   composePayload metadata >>=
+   mapM decrypt >>=
+   adjust metadata
+  where
+    checkPresenceOfMethod value =
+        let presentPair (TxMetaText k, TxMetaText v) =
+                k == cip83EncryptMethodKey && v == cip83EncryptPayloadValue
+            presentPair _ = False
+        in case value of
+            TxMetaMap list -> not (any presentPair list)
+            _ -> True
+    getEncryptedPayload value =
+        let presentPair (TxMetaText k, TxMetaList _) =
+                k == cip83EncryptPayloadKey
+            presentPair _ = False
+        in case value of
+            TxMetaMap list -> snd <$> filter presentPair list
+            _ -> []
+    extractTxt (TxMetaText txt) = txt
+    extractTxt _ =
+        error "TxMetaText is expected"
```

Comment:
🤔 I have the feeling this call to `error` will hit when given malformed metadata of this kind:

```hs
example = Cardano.TxMetadata $
Map.fromList
[ ( 674
  , Cardano.TxMetaMap
    [ ( Cardano.TxMetaText "enc"
      , Cardano.TxMetaText "basic"
      )
    , ( Cardano.TxMetaText "msg"
      , Cardano.TxMetaList
        [ Cardano.TxMetaText "Fm/+xoZBA24yp8Vz548NAg=="
        , Cardano.TxMetaNumber 123 -- ouch
        ]
      )
    ]
  )
]
```

Can you convert this `error` into `Left Err…` instead?

In order to keep the code transparent, I recommend using `Maybe` as a monad. (You can also use `Either e`, but that is more cumbersome as you need to convert error types which is not worth doing here.) This technique often applies when you have many nested pattern matches — the combination of "validate the data with `filter` and then do the match" usually doesn't scale very well. (You seem to use this combination often, I recommend to not to that, but instead get accustomed to "pattern match as validation" instead.)

Here is one way of how you can do it:

```hs
getEncryptedPayload (TxMetaMap list) = do
    (_, TxMetaList xs) <- find isCip83EncryptPayload list
    txts <- for xs matchTxMetaText
    pure $ foldl' T.append T.empty txs
getEncryptedPayload _ =
    Nothing

matchTxMetaText (TxMetaText t) = Just t
matchTxMetaText _ = Nothing

isCip83EncryptPayload (TxMetaText k, TxMetaList _) =
    k == cip83EncryptPayloadKey
isCip83EncryptPayload _ =
    False
```

Note that I like repeating the function name `isCip83EncryptPayload` for each case — this is intentional, and very useful for equational reasoning; for example, this definition can also be read *in reverse*.

Reply:
good idea, thanks. added handling of this case plus unit test checking this